WalletWasabi/WabiSabi/Client/Banning/CoinPrison.cs

```diff
@@ -0,0 +1,116 @@
+using Newtonsoft.Json;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using WalletWasabi.Blockchain.TransactionOutputs;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+using WalletWasabi.Wallets;
+
+namespace WalletWasabi.WabiSabi.Client.Banning;
+
+public class CoinPrison
+{
+	public CoinPrison(string filePath)
+	{
+		FilePath = filePath;
+	}
+
+	public List<PrisonedCoinRecord> BannedCoins { get; set; } = new();
+	public string FilePath { get; set; }
+	private object Lock { get; set; } = new();
+
+	public bool TryGetOrRemoveBannedCoin(SmartCoin coin, DateTimeOffset banDeadlineTime, [NotNullWhen(true)] out DateTimeOffset? bannedUntil)
+	{
+		bannedUntil = null;
+		if (BannedCoins.SingleOrDefault(record => record.Outpoint == coin.Outpoint) is { } record)
+		{
+			if (banDeadlineTime < record.BannedUntil)
+			{
+				bannedUntil = record.BannedUntil;
+				return true;
+			}
+			RemoveBannedCoin(coin);
+		}
+		return false;
+	}
+
+	public void Ban(SmartCoin coin, DateTimeOffset until)
+	{
+		if (BannedCoins.SingleOrDefault(record => record.Outpoint == coin.Outpoint) is { } record)
+		{
+			if (record.BannedUntil >= until)
+			{
+				return;
+			}
+		}
+		BannedCoins.Add(new(coin.Outpoint, until));
+		coin.BannedUntilUtc = until;
+		ToFile();
+	}
+
+	private void RemoveBannedCoin(SmartCoin coin)
+	{
+		var recordToRemove = BannedCoins.SingleOrDefault(record => coin.Outpoint == record.Outpoint);
+		if (recordToRemove == null)
+		{
+			Logger.LogError($"Tried to remove {nameof(coin)} from {nameof(BannedCoins)}, but {nameof(coin)} was null.");
+			return;
+		}
+		BannedCoins.Remove(recordToRemove);
+		coin.BannedUntilUtc = null;
+		ToFile();
+	}
+
+	private void ToFile()
+	{
+		lock (Lock)
+		{
+			if (string.IsNullOrWhiteSpace(FilePath))
+			{
+				return;
+			}
+
+			IoHelpers.EnsureFileExists(FilePath);
```

Comment:
`File.WriteAllText(...)` will make sure the file exists or if it exists it will overwrite it.

Do we need this line?

Reply:
With the usage of WriteAllText, not really. Removing.